### PR TITLE
Consolidate compute controller

### PR DESCRIPTION
**Pull request by teskje**
This PR refactors the compute controller to prepare it for the introduction of compute controller commands that record all state changes. The key change is that `Instance` and `ActiveInstance` are made private, so clients cannot change their state directly anymore. Instead, all state changes must now go through `ComputeController` and `ActiveComputeController`, allowing us to add the command boundary there in a follow-up PR.

### Motivation

  * This PR adds a known-desirable feature.

Part of #13222.

### Tips for reviewer

Look at the commits separately!

To get an idea of how introducing the controller commands will look like, check out https://github.com/MaterializeInc/materialize/pull/14682. This draft introduces compute controller commands, but only for per-instance controllers. It won't get merged but I will use the same approach for the whole compute controller.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/compute-client/src/controller.rs [hunk 1]
```diff
@@ -369,27 +372,31 @@ where
...
-    /// When the method returns, the caller should call [`ActiveComputeController::process`] with
-    /// the returned compute instance ID, to process the ready message.
+    /// When the method returns, the caller should call [`ActiveComputeController::process`] to
+    /// process the ready message.
     ///
     /// This method is cancellation safe.
-    pub async fn ready(&mut self) -> ComputeInstanceId {
+    pub async fn ready(&mut self) {
+        if self.stashed_response.is_some() {
+            // We still have a response stashed, which we are immediately ready to process.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:412]: is here a `return` missing? Or maybe better an assert?

File: src/compute-client/src/controller.rs [hunk 2]
```diff
@@ -489,6 +533,106 @@ where
         Ok(())
     }
 
+    /// Create and maintain the described dataflows, and initialize state for their output.
+    ///
+    /// This method creates dataflows whose inputs are still readable at the dataflow `as_of`
+    /// frontier, and initializes the outputs as readable from that frontier onward.
+    /// It installs read dependencies from the outputs to the inputs, so that the input read
+    /// capabilities will be held back to the output read capabilities, ensuring that we are
+    /// always able to return to a state that can serve the output read capabilities.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:540]: Can you add in the docstring in which cases this method can fail?

File: src/compute-client/src/controller.rs [hunk 3]
```diff
@@ -489,6 +533,106 @@ where
...
+        instance_id: ComputeInstanceId,
+        dataflows: Vec<DataflowDescription<crate::plan::Plan<T>, (), T>>,
+    ) -> Result<(), ComputeError> {
+        self.instance(instance_id)?
+            .create_dataflows(dataflows)
+            .await
+    }
+
+    /// Drop the read capability for the given collections and allow their resources to be
+    /// reclaimed.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:552]: (same here) Can you add in the docstring in which cases this method can fail?

I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call `.unwrap()` or not

File: src/compute-client/src/controller.rs [hunk 4]
```diff
@@ -489,6 +533,106 @@ where
...
+        &mut self,
+        instance_id: ComputeInstanceId,
+        collection_ids: Vec<GlobalId>,
+    ) -> Result<(), ComputeError> {
+        self.instance(instance_id)?
+            .drop_collections(collection_ids)
+            .await
+    }
+
+    /// Initiate a peek request for the contents of the given collection at `timestamp`.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:563]: (same here) Can you add in the docstring in which cases this method can fail?

I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call `.unwrap()` or not

File: src/compute-client/src/controller.rs [hunk 5]
```diff
@@ -489,6 +533,106 @@ where
...
+    ///
+    /// Canceling a peek is best effort. The caller may see any of the following
+    /// after canceling a peek request:
+    ///
+    ///   * A `PeekResponse::Rows` indicating that the cancellation request did
+    ///    not take effect in time and the query succeeded.
+    ///
+    ///   * A `PeekResponse::Canceled` affirming that the peek was canceled.
+    ///
+    ///   * No `PeekResponse` at all.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:598]: (same here) Can you add in the docstring in which cases this method can fail?

I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call `.unwrap()` or not

File: src/compute-client/src/controller.rs [hunk 6]
```diff
@@ -489,6 +533,106 @@ where
...
+    }
+
+    /// Assign a read policy to specific identifiers.
+    ///
+    /// The policies are assigned in the order presented, and repeated identifiers should
+    /// conclude with the last policy. Changing a policy will immediately downgrade the read
+    /// capability if appropriate, but it will not "recover" the read capability if the prior
+    /// capability is already ahead of it.
+    ///
+    /// Identifiers not present in `policies` retain their existing read policies.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:614]: (same here) Can you add in the docstring in which cases this method can fail?

I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call `.unwrap()` or not

File: src/compute-client/src/controller.rs [hunk 7]
```diff
@@ -489,6 +533,106 @@ where
...
+    pub async fn set_read_policy(
+        &mut self,
+        instance_id: ComputeInstanceId,
+        policies: Vec<(GlobalId, ReadPolicy<T>)>,
+    ) -> Result<(), ComputeError> {
+        self.instance(instance_id)?.set_read_policy(policies).await
+    }
+
+    /// Update the max size in bytes of any result.
```
Comment (lluki) [inline src/compute-client/src/controller.rs:623]: (same here) Can you add in the docstring in which cases this method can fail?

I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call `.unwrap()` or not

Btw, how does this actually work? What do we do if the result would exceed that size?

**Review by lluki — APPROVED** [review]

File: src/compute-client/src/controller.rs (hunk 1 above)
Comment (teskje) [reply src/compute-client/src/controller.rs:412]: Oof, good catch!

**Comment by teskje** [comment]
Regarding adding more documentation on the scenarios in which the individual controller methods can fail, I think this should be done in a separate PR. I think once we have the compute controller commands in place, with a separate validation step for each command, this will be much more straightforward than it currently is.

There is still the danger that such documentation can become wrong/outdated easily, as it documents behavior that is not local to the documented function. Ideally the error type would reflect all possible error scenarios, but of course in our case the error type is a superset of what can possibly happen and is not very useful for finding out what can actually happen. So I agree that we should have this documented instead.

> I think it's in particular important to understand if the only failure mode is invalid id or if there are also other cases, that gives the caller some confidence whether to call .unwrap() or not

One other failure mode that probably applies to most of the controller's methods is that communication with the storage controller can return an error. This is true, for example, for any method that can cause changes to storage collection frontiers. I have no idea what the failure modes of the storage controller are, though.

**Pull request merged**